### PR TITLE
WIP/Proposal - Cluster name-prefixed load balancers

### PR DESCRIPTION
**Pull request by Preskton**
### What type of PR is this?

/kind api-change

... since this changes the contract for how we name load balancers when created by the cloud provider.

### What this PR does / why we need it

As an operator, I want the load balancer names that are created via Kubernetes to have semi-predictable names so that I can do things like limit the cluster's service principal role IAM to only affect resources prefixed with the cluster name.

This also replaces a method that is now deprecated in the base cloud provider implementation, `DefaultLoadBalancerName`, ref: https://github.com/kubernetes/cloud-provider/blob/02ef7372f6c090f5a90430f1b178b38f1fba2b1e/cloud.go#L82.

If anything goes badly, we fall back to the prior implementation (`a` + the service's UID).

### Which issue(s) this PR fixes

No issue has been specifically raised for this. Should I do this as part of the process?

### Special notes for your reviewer

This is my first PR into `cloud-provider-aws`. I'm not even sure that I've submitted it to the right place. Happy to adjust how I've done this, test coverage, etc. - feedback much appreciated! :)

I'll likely add some additional coverage around the edge cases of:

- greater than 32 chars
- special characters - although my guess this is not possible based on validation that's already happening upstream in AWS on cluster name and k8s on service name.

### Does this PR introduce a user-facing change?

```release-note
Load Balancers: When creating a load balancer for a `Service` of type `LoadBalancer`, the name of the new load balancer will now follow the format: `$clusterName-$serviceName`, limited to 32 characters.
```

### CODE REVIEW

**Comment by Preskton** [comment]
The 32 character limit restriction is pretty intense: someone could pretty easily reduce the set of load balancers they could create by creating a cluster with a fairly long name.

Would this make more sense if we tagged the xLB in AWS with `kubernetes.io/cluster` with a value of the cluster name? I think this would serve my purpose of being able to restrict IAM with `iam:ResourceTag` filters.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -45,7 +46,7 @@ import (
 	gcfg "gopkg.in/gcfg.v1"
 	"k8s.io/klog"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (jaypipes) [inline pkg/cloudprovider/providers/aws/aws.go]: technically, this isn't necessary. the package name is already "v1". but, it doesn't hurt anything I suppose :)

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -3663,8 +3664,18 @@ func (c *Cloud) GetLoadBalancer(ctx context.Context, clusterName string, service
 
 // GetLoadBalancerName is an implementation of LoadBalancer.GetLoadBalancerName
 func (c *Cloud) GetLoadBalancerName(ctx context.Context, clusterName string, service *v1.Service) string {
-	// TODO: replace DefaultLoadBalancerName to generate more meaningful loadbalancer names.
-	return cloudprovider.DefaultLoadBalancerName(service)
+	// Ref: https://docs.aws.amazon.com/elasticloadbalancing/2012-06-01/APIReference/API_CreateLoadBalancer.html
+	// unique for the region / max 32 chars / alphanumeric plus hyphen / can't start or end with hyphen
+	reg := regexp.MustCompile(`[^a-zA-Z0-9\-]+`)
```
Comment (jaypipes) [inline pkg/cloudprovider/providers/aws/aws.go]: That regex actually would allow a hyphen at the start or end of the name.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 3]
```diff
@@ -3663,8 +3664,18 @@ func (c *Cloud) GetLoadBalancer(ctx context.Context, clusterName string, service
...
+
+	ret := fmt.Sprintf("%s-%s", clusterName, string(service.UID))
+	ret = reg.ReplaceAllString(ret, "")
+
+	if len(ret) > 32 {
+		ret = ret[:32]
```
Comment (jaypipes) [inline pkg/cloudprovider/providers/aws/aws.go]: Any cluster with a name 31 characters long will result in the ELB name for all LoadBalancers in the cluster being the same. I'm pretty sure that's not what you want, right?

Comment (Preskton) [reply pkg/cloudprovider/providers/aws/aws.go]: Yup, unfortunately, that's the char limit for LB names. :(

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (Preskton) [reply pkg/cloudprovider/providers/aws/aws.go]: Must be my import settings - will clean up.
